streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamWindowAggregate.java

```diff
@@ -80,22 +109,54 @@ public void enableSendingOldValues() {
+        private Sensor emitFinalLatencySensor;
         private long observedStreamTime = ConsumerRecord.NO_TIMESTAMP;
+        private long lastEmitCloseTime = ConsumerRecord.NO_TIMESTAMP;
+        private InternalProcessorContext<Windowed<KIn>, Change<VAgg>> internalProcessorContext;
+        private final TimeTracker timeTracker = new TimeTracker();
+        private final Time time = Time.SYSTEM;
 
         @Override
         public void init(final ProcessorContext<Windowed<KIn>, Change<VAgg>> context) {
             super.init(context);
-            final InternalProcessorContext<Windowed<KIn>, Change<VAgg>> internalProcessorContext =
-                (InternalProcessorContext<Windowed<KIn>, Change<VAgg>>) context;
+            internalProcessorContext = (InternalProcessorContext<Windowed<KIn>, Change<VAgg>>) context;
             final StreamsMetricsImpl metrics = internalProcessorContext.metrics();
             final String threadId = Thread.currentThread().getName();
             droppedRecordsSensor = droppedRecordsSensor(threadId, context.taskId().toString(), metrics);
+            emittedRecordsSensor = emittedRecordsSensor(threadId, context.taskId().toString(), metrics);
+            emitFinalLatencySensor = emitFinalLatencySensor(threadId, context.taskId().toString(), metrics);
             windowStore = context.getStateStore(storeName);
-            tupleForwarder = new TimestampedTupleForwarder<>(
-                windowStore,
-                context,
-                new TimestampedCacheFlushListener<>(context),
-                sendOldValues);
+
+            if (emitStrategy.type() == StrategyType.ON_WINDOW_CLOSE) {
+                // Don't set flush lister which emit cache results
+                tupleForwarder = new TimestampedTupleForwarder<>(
+                    windowStore,
+                    context,
+                    sendOldValues);
+            } else {
+                tupleForwarder = new TimestampedTupleForwarder<>(
+                    windowStore,
+                    context,
+                    new TimestampedCacheFlushListener<>(context),
+                    sendOldValues);
+            }
+
+            log.info("EmitStrategy=" + emitStrategy.type());
+            // Restore last emit close time for ON_WINDOW_CLOSE strategy
+            if (emitStrategy.type() == StrategyType.ON_WINDOW_CLOSE) {
+                final Long lastEmitTime = internalProcessorContext.processorMetadataForKey(storeName);
+                if (lastEmitTime != null) {
+                    lastEmitCloseTime = lastEmitTime;
+                }
+                final long emitInterval = StreamsConfig.InternalConfig.getLong(
+                    context.appConfigs(),
+                    EMIT_INTERVAL_MS_KSTREAMS_WINDOWED_AGGREGATION,
+                    1000L
+                );
+                timeTracker.setEmitInterval(emitInterval);
+                log.info("EmitInterval=" + emitInterval);
```

Comment:
Similar to previous comment, I feel this won't be called many time (just once when creating the processor?). And the config won't be printed out when we print all configs since this is an internal config. This will be convenient for debugging.

Reply:
It's not about spamming but it's a question if it's useful? Also seems to leak an internal config that we might want to keep hidden?